geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/FeatureSpecificReader.scala

```diff
@@ -96,9 +94,12 @@ class FeatureSpecificReader(oldType: SimpleFeatureType, newType: SimpleFeatureTy
   lazy val v1fieldreaders = buildFieldReaders(Version1Deserializer)
   lazy val v2fieldreaders = buildFieldReaders(Version2Deserializer)
 
-  def read(reuse: AvroSimpleFeature, in: Decoder): AvroSimpleFeature = {
-    // Read the version first and choose the proper deserializer
+  def defaultRead(reuse: AvroSimpleFeature, in: Decoder): AvroSimpleFeature = {
+    // read and store the version
     val serializationVersion = in.readInt()
+    AvroSimpleFeatureDecodingsCache.getAbstractReader.version = serializationVersion
```

Comment:
I agree that this isn't great.  However, it is not static - there is one instance per thread.  And, at least for Kryo, there is a hard reference through the "val encodings".

Initially, I was passing the version as a parameter to the DatumReader but when debugging I noticed a lot of boxing and unboxing.  And, the version is rarely used which is why I removed it.

A cleaner solution is to specify the version at the time that the attribute decodings are built and to cache the decodings for multiple versions.  I will make this change.